metricshub-agent/src/main/java/org/sentrysoftware/metricshub/agent/helper/ConnectorVariablesLibraryParser.java

```diff
@@ -0,0 +1,288 @@
+import org.sentrysoftware.metricshub.engine.connector.parser.ConnectorParser;
+
+/**
+ * Utility class for parsing connectors with variables and creating a map of custom connectors.
+ * This class provides functionality to visit YAML files in a specified directory, read connector data,
+ * and create Connector objects based on the parsed data. The resulting connectors are stored in a map
+ * with the connector ID as the key and the corresponding Connector object as the value.
+ * <p>
+ * The parsing process involves checking for YAML files, validating whether the YAML structure defines a
+ * final Connector (with a displayName section), and using ConnectorParser to parse the YAML file and create
+ * Connector objects.
+ * </p>
+ */
+@Slf4j
+public class ConnectorVariablesLibraryParser {
+
+	/**
+	 * This inner class allows to visit the files contained within the Yaml directory
+	 */
+	private static class ConnectorFileVisitor extends SimpleFileVisitor<Path> {
+
+		private final Map<String, AdditionalConnector> additionalConnectorConfig;
+		private final AdditionalConnectorsParsingResult connectorsParsingResult = new AdditionalConnectorsParsingResult();
+
+		public AdditionalConnectorsParsingResult getConnectorsParsingResult() {
+			return connectorsParsingResult;
+		}
+
+		ConnectorFileVisitor(final Map<String, AdditionalConnector> additionalConnectorConfig) {
+			this.additionalConnectorConfig = additionalConnectorConfig;
+		}
+
+		@Override
+		public FileVisitResult visitFile(Path path, BasicFileAttributes attrs) throws IOException {
+			// Skip this path if it is a directory or not a YAML file
+			if (Files.isDirectory(path) || !isYamlFile(path.toFile().getName())) {
+				return FileVisitResult.CONTINUE;
+			}
+			final ObjectMapper yamlMapper = new YAMLMapper();
+			final JsonNode connectorNode = yamlMapper.readTree(path.toFile());
+			if (!isConnector(connectorNode)) {
+				return FileVisitResult.CONTINUE;
+			}
+
+			// Get the connector's file name
+			final String filename = path.getFileName().toString();
+			final String connectorId = filename.substring(0, filename.lastIndexOf('.'));
+
+			if (!connectorNode.toString().contains("${var::")) {
+				return FileVisitResult.CONTINUE;
+			}
+
+			// Normalize additionalConnectors
+			normalizeAdditionalConnectors();
+
+			// Removing all the configurations that are not using this connector.
+			final Map<String, AdditionalConnector> filteredConnectors = additionalConnectorConfig
+				.entrySet()
+				.stream()
+				.filter(entry -> connectorId.equalsIgnoreCase(entry.getValue().getUses()))
+				.collect(
+					Collectors.toMap(
+						Map.Entry::getKey, // Keep the original key
+						Map.Entry::getValue // Keep the original AdditionalConnectorConfig as the value
+					)
+				);
+
+			// Construct a variables map from the default connector variables.
+			final Map<String, String> defaultVariables = new HashMap<>(getDefaultConnectorVariables(connectorNode));
```

Comment:
What happens if the default variables are empty?

Reply:
`getDefaultConnectorVariables` returns a new HashMap if the default variables are empty. Consequently, there are two cases:
1. User defined default values: in this case, these values will be used.
2. User didn't define default values: in this case: ${var:: remain in the connector.